runtime/jvmti/heapify.cpp

```diff
@@ -0,0 +1,355 @@
+			J9JVMTIObjectMap exemplar;
+			initializeExemplar(walkState->walkThread, obj, &exemplar);
+			J9JVMTIObjectMap *entry = (J9JVMTIObjectMap*)hashTableFind(objectMap, &exemplar);
+			if (NULL != entry) {
+//				PORT_ACCESS_FROM_VMC(currentThread);
+//				j9tty_printf(PORTLIB, "<%p> Replacing %p with %p in slot %p\n", walkState->walkThread, obj, entry->heapObject, stackLocation);
+				*objectSlot = entry->heapObject;
+			}
+		}
+	}
+}
+
+static j9object_t
+allocateHeapObject(J9VMThread *currentThread, j9object_t object)
+{
+	j9object_t heapObject = NULL;
+	J9Class *objectClass = J9OBJECT_CLAZZ(currentThread, object);
+	UDATA classFlags = J9CLASS_FLAGS(objectClass);
+	J9MemoryManagerFunctions const * const mmFuncs = currentThread->javaVM->memoryManagerFunctions;
+#if 0
+	j9objectmonitor_t *lockEA = NULL;
+
+	if (LN_HAS_LOCKWORD(currentThread, object)) {
+		lockEA = J9OBJECT_MONITOR_EA(currentThread, object);
+	} else {
+		J9ObjectMonitor *objectMonitor = monitorTableAt(currentThread, object);
+		// Should probably just assert this is NULL
+		if (NULL != objectMonitor) {
+			lockEA = &(objectMonitor->alternateLockword);
+		}
+	}
+	if (NULL != lockEA) {
+		Assert_JVMTI_true(NULL == J9_LOAD_LOCKWORD(currentThread, lockEA));
+	}
+#endif
+	if (J9_ARE_ANY_BITS_SET(classFlags, J9AccClassArray)) {
+		U_32 size = J9INDEXABLEOBJECT_SIZE(currentThread, object);
+		heapObject = mmFuncs->J9AllocateIndexableObject(currentThread, objectClass, size, J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE);
+	} else {
+		heapObject = mmFuncs->J9AllocateObject(currentThread, objectClass, J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE);
+	}
+
+	return heapObject;
+}
+
+static j9object_t
+allocateHeapObjects(J9VMThread *currentThread, J9HashTable *objectMap)
+{
+	uint32_t objectCount = hashTableGetCount(objectMap);
+	J9JavaVM *vm = currentThread->javaVM;
+	J9Class *jlObject = J9VMJAVALANGOBJECT_OR_NULL(vm);
+	J9Class *arrayClass = jlObject->arrayClass;
+	j9object_t objectArray = vm->memoryManagerFunctions->J9AllocateIndexableObject(currentThread, arrayClass, objectCount, J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE);
+	if (NULL != objectArray) {
+		uint32_t index = 0;
+		J9HashTableState hashWalkState;
+		J9JVMTIObjectMap *entry = (J9JVMTIObjectMap*)hashTableStartDo(objectMap, &hashWalkState);
+		PUSH_OBJECT_IN_SPECIAL_FRAME(currentThread, objectArray);
+		while (NULL != entry) {
+			j9object_t heapObject = allocateHeapObject(currentThread, stackObjectFromMapEntry(entry));
+			if (NULL == heapObject) {
+				objectArray = NULL;
+				break;
+			}
+			objectArray = PEEK_OBJECT_IN_SPECIAL_FRAME(currentThread, 0);
+			J9JAVAARRAYOFOBJECT_STORE(currentThread, objectArray, index, heapObject);
+			index += 1;
+			entry = (J9JVMTIObjectMap*)hashTableNextDo(&hashWalkState);
+		}
+		DROP_OBJECT_IN_SPECIAL_FRAME(currentThread);
+	}
+	return objectArray;
+}
+
+static void
+copyObjectContents(J9VMThread *currentThread, J9JVMTIObjectMap *entry, J9HashTable *objectMap)
```

Comment:
Think about having this method within MM_ObjectAccessBarrierAPI, while keeping stack allocated specific parts (around objectIsStackAllocated())  here in smaller helper methods.

Reply:
To keep the code contained, I suppose I would actually pass a mapping function pointer (and perhaps an opaque userData field) rather than the table itself. The existing public APIs would remain unchanged (passing NULLs for the new fields to the more general helper).

Reply:
When you say 'this solution', do you mean what you have currently in the PR or generalized slow path (with references) copyObjectFields that would include this new logic? Yes, I prefer the latter.
If we later do realize we need a specialized fast implementation in MM_ObjectAccessBarrierAPI, we can add at that time.

Reply:
OK, I will go ahead with the parameterized copyObjectFields (and callers). As you say, it would be easy to specialize it again at a later date if it becomes necessary.